project-management/src/main/java/com/smartagilify/projectmanagement/services/wbsprogressreport/WBSProgressReportServiceImpl.java

```diff
@@ -0,0 +1,78 @@
+import com.smartagilify.projectmanagement.repositories.wbsprogressreport.WBSProgressReportRepository;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Service;
+
+import java.time.LocalDateTime;
+
+@Service
+public class WBSProgressReportServiceImpl extends
+        BaseService<WBSProgressReport, WBSProgressReportMapper, WBSProgressReportDTO> implements WBSProgressReportService {
+    private final WBSProgressReportRepository wbsProgressReportRepository;
+    private final PerformanceReportRepository performanceReportRepository;
+    private final WBSRepository wbsRepository;
+
+    public WBSProgressReportServiceImpl(JpaRepository<WBSProgressReport, Long> jpaRepository,
+                                        WBSProgressReportRepository wbsProgressReportRepository,
+                                        PerformanceReportRepository performanceReportRepository,
+                                        WBSRepository wbsRepository) {
+        super(jpaRepository);
+        this.wbsProgressReportRepository = wbsProgressReportRepository;
+        this.performanceReportRepository = performanceReportRepository;
+        this.wbsRepository = wbsRepository;
+    }
+
+    @Override
+    protected Class<WBSProgressReportMapper> getMapper() {
+        return WBSProgressReportMapper.class;
+    }
+
+    @Override
+    public WBSProgressReportDTO save(WBSProgressReportDTO dto, Long userId) {
+        WBSProgressReport entity = mapper.dto2Entity(dto);
+        PerformanceReport performanceReport = performanceReportRepository
+                .getReferenceById(dto.getPerformanceReportId());
+        WBS wbs = wbsRepository.getReferenceById(dto.getWbsId());
+        entity.setPerformanceReport(performanceReport);
+        entity.setWbs(wbs); entity.setState(EN_STATE.CREATED);
+        entity.setCreateById(userId); entity.setCreateDate(LocalDateTime.now());
+        return mapper.entity2Dto(wbsProgressReportRepository.save(entity));
+    }
+
+    @Override
+    public WBSProgressReportDTO update(WBSProgressReportDTO dto, Long userId) {
+        if(dto.getId() == null) throw new ApplicationException("This entity is not yet saved");
+        WBSProgressReport entity = mapper.dto2Entity(dto);
+        PerformanceReport performanceReport = performanceReportRepository
+                .getReferenceById(dto.getPerformanceReportId());
+        WBS wbs = wbsRepository.getReferenceById(dto.getWbsId());
+        entity.setPerformanceReport(performanceReport);
+        entity.setWbs(wbs); entity.setState(EN_STATE.UPDATED);
+        entity.setCreateById(userId); entity.setCreateDate(LocalDateTime.now());
+        return mapper.entity2Dto(wbsProgressReportRepository.save(entity));
+    }
+
+    @Override
+    public void softDelete(WBSProgressReportDTO dto, Long userId) {
+        if(dto.getId() == null) throw new ApplicationException("This entity is not yet saved");
```

Comment:
I think crating _static final_ field in `ErrorConstant ` class can be a good practice to store this kind of error message. isn't it?

Reply:
yes, we need to talk about where should we place the constants

Reply:
it cloud be like core